### ISSUE TITLE
Training with Batch

### ISSUE THREAD

**Issue by ShuaibZyx** [sail-sg/TreeMeshGPT#9]
Hello, I am very interested in your work. I am trying to modify my single-batch training code to include batch training, but I am encountering some issues. Could you provide a template with batch training so that I can troubleshoot the problem? Thank you for your help.

**Comment by splionar** [comment]
Hi @ShuaibZyx ,

To use batch size more than 1, you might want to familiarize with the collate function from PivotMesh codebase, which our codebase is inherited from, specifically in this line: https://github.com/whaohan/pivotmesh/blob/ed5652d7584e631cd400fc44e8fee285289e1482/trainer.py#L144. You might need to follow meshgpt_pytorch installation from PivotMesh, then comment out https://github.com/sail-sg/TreeMeshGPT/blob/35821582ac14090f2a9cb173e4ea5fa022ca760a/trainer.py#L144 as well as meshgpt_pytorch imports on the same file.

The collate function will apply necessary padding following on the longest sequence in a batch. Afterward, you need to adjust the following code snippet to make it work with batch size > 1.
https://github.com/sail-sg/TreeMeshGPT/blob/35821582ac14090f2a9cb173e4ea5fa022ca760a/model/treemeshgpt_train.py#L116-L125

This code snippet is used to truncate trailing [STOP] labels to make the sequence more efficient. Simple example of what it does is as follows:

```
Say the output sequence is the following:
Original Output:
[..., (new vertex), (new vertex), (STOP), (new vertex), (new vertex), (STOP), (STOP), (STOP)]

Adjusted Output with [EOS]:
[..., (new vertex), (new vertex), (STOP), (new vertex), (new vertex), (EOS)]

The [EOS] token replaces the first (boundary) token that appears after the last (new vertex).

The input sequence is the same, but truncated until [EOS].

```

**Comment by ShuaibZyx** [comment]
Thank you for your answer @splionar, as you said, I applied the collate function in meshgpt_pytorch:
```
def custom_collate(data, pad_id = -1):
    is_dict = isinstance(first(data), dict)
    if is_dict:
        keys = first(data).keys()
        data = [d.values() for d in data]
    output = []
    for datum in zip(*data):
        if is_tensor(first(datum)):
            datum = pad_sequence(datum, batch_first = True, padding_value = pad_id)
        else:
            datum = list(datum)
        output.append(datum)
    output = tuple(output)
    if is_dict:
        output = dict(zip(keys, output))
    return output
```
and modified it slightly to:
```
def collate_model_batch(self, ds: List[Dict[str, torch.Tensor]]) -> Dict[str, torch.Tensor]:
    keys = ds[0].keys()
    values = zip(*(d.values() for d in ds))
    collated = {}
    for key, value in zip(keys, values):
        if torch.is_tensor(next(iter(value))):
            collated[key] = pad_sequence(value, batch_first=True, padding_value=-2)
        else:
            collated[key] = list(value)
    return collated
```
, where the -2 is the set end mark (EOS), and -1 is the one mentioned in your code (STOP).
After that, I set the (STOP) you marked at the end of the sequence to -2.
```
# construct data
for i in range(len(edges)):
    gt_ind[i].insert(0, edges[i][0][1])
    gt_ind[i].insert(0, edges[i][0][0])
    edges[i].insert(0, [edges[i][0][0], -1])
    edges[i].insert(0, [-1, -1])

gt_combine = [elem for sublist in gt_ind for elem in sublist]
edges_combine = [elem for sublist in edges for elem in sublist]

gt_combine.append(-2)  # eos
edges_combine.append([-2, -2])
```
After processing, the sequence should be:

[..., (new vertex), (new vertex), (STOP), (new vertex), (new vertex), (EOS), (EOS), (EOS)]

I wrote a method to simulate the truncation effect of your single batch and set the end (STOP) and padding uniformly to -2, as shown:
```
def _construct_mask(self, gt_ind, edges):
    B, BML = gt_ind.shape
    gt_ind[gt_ind == -2] = -1
    valid_mask = gt_ind != -1
    indices = (
        valid_mask.cumsum(dim=1).eq(valid_mask.sum(dim=1, keepdim=True))
        & valid_mask
    )
    last_indices = indices.float().argmax(dim=1) + 1
    last_indices = torch.clamp(last_indices, max=BML - 1)
    range_tensor = torch.arange(BML, device=self.device).unsqueeze(0).expand(B, BML)
    mask = range_tensor <= last_indices.unsqueeze(1)

    padding_mask = range_tensor >= last_indices.unsqueeze(1)
    gt_ind[padding_mask] = -2
    edges[padding_mask] = -2
    return mask
```
Finally, I made the following modifications to the main code:
```
self.eos_emb = nn.Parameter(torch.randn(dim))

gt[gt_ind == -2] = 2**self.quant_bit + 1

B, _, C = vertices.shape
 _, BML = gt_ind.shape

edge_sos_type = (edges == torch.tensor([-1, -1], device=self.device)).sum(-1)
edge_eos_type = (edges == torch.tensor([-2, -2], device=self.device)).sum(-1)
edges_emb[edge_sos_type == 2] = self.sos_emb
edges_emb[edge_eos_type == 2] = self.eos_emb
```
I ran the experiment, and the train loss decreased very quickly, while the val loss decreased slightly and then overfit.
![Image](https://github.com/user-attachments/assets/18bb265e-0906-4735-aa9c-a36e61beaa36)

![Image](https://github.com/user-attachments/assets/1759f2e5-c278-46da-84b7-a90f266c692b)
I am confused as to whether my approach to handling the sequence is incorrect.
Looking forward to your guidance and thank you.

**Comment by splionar** [comment]
Hi @ShuaibZyx ,

Just a quick check, have you applied the collate function to the val dataloader: https://github.com/sail-sg/TreeMeshGPT/blob/35821582ac14090f2a9cb173e4ea5fa022ca760a/trainer.py#L161?

**Comment by ShuaibZyx** [comment]
Hi @splionar ,
Yes, it is applied. I checked the sequence elements, and they are consistent with the expected results.
```
def val_dataloader(self) -> DataLoader:
        return DataLoader(
            self.val_dataset,
            self.batch_size,
            shuffle=False,
            drop_last=True,
            collate_fn=self.collate_model_batch,
            num_workers=4,
            persistent_workers=True,
        )
```

**Comment by splionar** [comment]
@ShuaibZyx I am not sure why it overfits to the training data in your case. Have you kept python train_create_pkl.py running in a seperate process to keep the data augmentation?

**Comment by ShuaibZyx** [comment]
Regardless of overfitting, I would like to ask you if my modification method is the correct approach. @splionar 😋

**Comment by ShuaibZyx** [comment]
Hello, sorry to bother you again. I would like to ask, if a special value such as [PAD] is filled in after vertex (gt) in [EOS], how should the corresponding edge be filled (a single batch is truncated)? Or, what should be set for the fill part in the edges_emb vector? @splionar Thank you!

**Comment by ForrestPi** [comment]
@ShuaibZyx  Hi, how to Training with Batch, could you solve it?  could you share your code?
